src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -80,11 +85,221 @@ static ShardPlacement * ShardPlacementOnGroup(uint64 shardId, int groupId);
+		ereport(WARNING, (errcode(ERRCODE_CONNECTION_FAILURE),
+						  errmsg("cannot get the disk space statistics for node %s:%d",
+								 nodeName, nodePort)));
+		return false;
+	}
+
+	if (!IsResponseOK(result) || PQntuples(result) != 1)
+	{
+		ereport(WARNING, (errcode(ERRCODE_CONNECTION_FAILURE),
+						  errmsg("cannot get the disk space statistics for node %s:%d",
+								 nodeName, nodePort)));
+
+		PQclear(result);
+		ClearResults(connection, raiseErrors);
+
+		return false;
+	}
+
+	char *availableBytesString = PQgetvalue(result, 0, 0);
+	char *totalBytesString = PQgetvalue(result, 0, 1);
+
+	*availableBytes = SafeStringToUint64(availableBytesString);
+	*totalBytes = SafeStringToUint64(totalBytesString);
+
+	PQclear(result);
+	ClearResults(connection, raiseErrors);
+
+	return true;
+}
+
+
+/*
+ * citus_local_disk_space_stats returns total disk space and available disk
+ * space for the disk that contains PGDATA.
+ */
+Datum
+citus_local_disk_space_stats(PG_FUNCTION_ARGS)
+{
+	uint64 availableBytes = 0;
+	uint64 totalBytes = 0;
+
+	if (GetLocalDiskSpaceStats(&availableBytes, &totalBytes))
+	{
+		ereport(WARNING, (errmsg("could not get disk space")));
+	}
+
+	TupleDesc tupleDescriptor = NULL;
+	Datum values[TABLE_METADATA_FIELDS];
+	bool isNulls[TABLE_METADATA_FIELDS];
+
+	TypeFuncClass resultTypeClass = get_call_result_type(fcinfo, NULL,
+														 &tupleDescriptor);
+	if (resultTypeClass != TYPEFUNC_COMPOSITE)
+	{
+		ereport(ERROR, (errmsg("return type must be a row type")));
+	}
+
+	/* form heap tuple for local disk space statistics */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	values[0] = UInt64GetDatum(availableBytes);
+	values[1] = UInt64GetDatum(totalBytes);
+
+	HeapTuple diskSpaceTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	PG_RETURN_DATUM(HeapTupleGetDatum(diskSpaceTuple));
+}
+
+
+/*
+ * GetLocalDiskSpaceStats returns total and available disk space for the disk containing
+ * PGDATA (not considering tablespaces, quota).
+ */
+static bool
+GetLocalDiskSpaceStats(uint64 *availableBytes, uint64 *totalBytes)
+{
+	struct statvfs buffer;
+	if (statvfs(DataDir, &buffer) != 0)
+	{
+		return false;
+	}
+
+	*availableBytes = buffer.f_bfree * buffer.f_frsize;
+	*totalBytes = buffer.f_blocks * buffer.f_frsize;
+
+	return true;
+}
+
+
+/*
+ * citus_database_size is a UDF that returns the sum of the database sizes
+ * across all nodes.
+ */
+Datum
+citus_database_size(PG_FUNCTION_ARGS)
+{
+	Datum databaseNameDatum = PG_GETARG_DATUM(0);
+	Name databaseName = DatumGetName(databaseNameDatum);
+	char *databaseNameString = NameStr(*databaseName);
+
+	CheckCitusVersion(ERROR);
+
+	Datum databaseSizeDatum = DirectFunctionCall1(pg_database_size_name,
+												  databaseNameDatum);
+
+	uint64 databaseSize = Int64GetDatum(databaseSizeDatum);
+
+	/* we calculated the coordinator database size above, so only query workers */
+	List *workerNodeList = ActiveReadableNonCoordinatorNodeList();
```

Comment:
what if the coordinator is added as `master_add_node('xx', port)`;  where we don't give groupid 0? Then we will count it twice?

Reply:
what was the reason we are doing local and remote separately? Can't we just get all the nodes in this line so that it can connect to localhost as well?

Reply:
Coordinator is not always in the metadata. It should never be added with groupid > 0, that'll lead to frequent deadlocks.

Reply:
Alright, I guess we already have an issue for this now: #4428